zokrates_ast/src/ir/serialize.rs

```diff
@@ -58,33 +63,195 @@ impl<
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum SectionType {
+    Arguments = 1,
+    Constraints = 2,
+    Solvers = 3,
+}
+
+impl TryFrom<u32> for SectionType {
+    type Error = String;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(SectionType::Arguments),
+            2 => Ok(SectionType::Constraints),
+            3 => Ok(SectionType::Solvers),
+            _ => Err("invalid section type".to_string()),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Section {
+    pub ty: SectionType,
+    pub offset: u64,
+    pub length: u64,
+}
+
+impl Section {
+    pub fn new(ty: SectionType) -> Self {
+        Self {
+            ty,
+            offset: 0,
+            length: 0,
+        }
+    }
+
+    pub fn set_offset(&mut self, offset: u64) {
+        self.offset = offset;
+    }
+
+    pub fn set_length(&mut self, length: u64) {
+        self.length = length;
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ProgHeader {
+    pub magic: [u8; 4],
+    pub version: [u8; 4],
+    pub curve_id: [u8; 4],
+    pub constraint_count: u32,
+    pub return_count: u32,
+    pub sections: Vec<Section>,
+}
+
+impl ProgHeader {
+    pub fn write<W: Write>(&self, mut w: W) -> std::io::Result<()> {
+        w.write_all(&self.magic)?;
+        w.write_all(&self.version)?;
+        w.write_all(&self.curve_id)?;
+        w.write_u32::<LittleEndian>(self.constraint_count)?;
```

Comment:
I included this in the header so we don't have to `collect` the `Prog` to read the constraint count